Python/ceval.c

```diff
@@ -3118,6 +3125,20 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, InterpreterFrame *frame, int thr
             DISPATCH();
         }
 
+        TARGET(COPY_FREE_VARS) {
+            /* Copy closure variables to free variables */
+            PyCodeObject *co = frame->f_code;
+            PyObject *closure = frame->f_func->func_closure;
```

Comment:
Although really unlikely, this introduces a small race, doesn't it?  Code elsewhere, in another thread, could swap out the function's `func_closure`.

Reply:
No. The race was already there.

Reply:
Let me elaborate. Unless there is a mutex, the change to __closure__ from one thread can be seen by another at any time.

The callee might see the old or new closure, but that was true before.
What cannot happen (and this is still true) is that the callee cannot see a change in the closure.

Reply:
I'm talking about the race from the time the frame gets initialized to the time this instruction is executed.  That did not exist before.  Again, it's extremely unlikely, but it *is* similar to the case we accommodate in `PyFrame_FastToLocals()`, etc.

Reply:
What do you mean by "frame gets initialized"? `COPY_FREE_VARS` is part of the initialization of the frame.

Reply:
Yeah, but that is executed after the eval loop has started (and the GIL possibly released).

Reply:
I still don't see a race condition here.

If you are worried about other code being able to see a partly created frame, that can't happen either.
 The only way that the free variables of a frame are visible is via the `f_locals` attribute and `_PyFrame_FastToLocalsWithError()` checks to see if the free variables need be initialized.

Reply:
Thanks for clarifying.